### PR TITLE
Add support for numeric QR codes

### PR DESCRIPTION
**Pull request by toncid**
Numeric QR codes can now be enabled with an extra initialization parameter:

```ruby
qr = RQRCode::QRCode.new( '1234567890', :size => 2, :level => :m, :mode => :number )
```

### CODE REVIEW

**Comment by toncid** [comment]
Fixes #33.

File: README.md [hunk 1]
```diff
@@ -140,6 +144,17 @@ qr.modules.each do |row|
 end
 ```
 
+## Specifying QR code mode
+
+Sometimes you may want to specify the QR code mode explicitly. 
+
+It is done voia the `mode` option. Allowed values are: `number`, `alphanumeric`, `8bit_byte` and `kanji`.
```
Comment (bjornblomqvist) [inline README.md]: should it maybe be "via the mode"? Also we don't support kanji, so maybe leave it out for now ?

File: lib/rqrcode/qrcode/qr_code.rb [hunk 2]
```diff
@@ -18,6 +18,13 @@ module RQRCode #:nodoc:
     :mode_kanji         => 1 << 3
   }
 
+  QRMODE_NAME = {
+    :number        => :mode_number,
+    :alphanumeric  => :mode_alpha_numk,
+    :'8bit_byte'   => :mode_8bit_byte,
```
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: Maybe just use :byte for the 8bit mode? Quoted symbols are kind of unpleasant to write.

File: lib/rqrcode/qrcode/qr_code.rb [hunk 3]
```diff
@@ -18,6 +18,13 @@ module RQRCode #:nodoc:
...
+    :kanji         => :mode_kanji
```
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: What do say about leaving kanji out from the list of modes. Its not supported so i am thinking i might just lead to misunderstandings.

File: lib/rqrcode/qrcode/qr_code.rb (hunk 2 above)
Comment (toncid) [inline lib/rqrcode/qrcode/qr_code.rb]: Or maybe both or inverse:

```ruby
QRMODE_NAME = {
  :number        => :mode_number,
  :alphanumeric  => :mode_alpha_numk,
  :'8bit_byte'   => :mode_8bit_byte,
  :byte          => :mode_8bit_byte,
  :byte_8bit     => :mode_8bit_byte,
  :kanji         => :mode_kanji
}
```

File: lib/rqrcode/qrcode/qr_code.rb [hunk 4]
```diff
@@ -117,7 +129,8 @@ def initialize( string, *args )
 
       @data                 = string
 
-      mode                  = QRAlphanumeric.valid_data?( @data ) ? :mode_alpha_numk : :mode_8bit_byte
+      mode                  = QRMODE_NAME[(options[:mode] || '').to_sym]
```
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: to_sym on an empty string does not work in 1.8.

File: README.md (hunk 1 above)
Comment (toncid) [inline README.md]: Yes, that's a typo. I agree regarding kanji.

File: lib/rqrcode/qrcode/qr_code.rb [hunk 5]
```diff
@@ -117,7 +129,8 @@ def initialize( string, *args )
...
+      mode                  ||= QRAlphanumeric.valid_data?( @data ) ? :mode_alpha_numk : :mode_8bit_byte  # deprecate?
```
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: What do you mean by deprecate?

File: lib/rqrcode/qrcode/qr_code.rb (hunk 3 above)
Comment (toncid) [inline lib/rqrcode/qrcode/qr_code.rb]: Agreed (for now).

File: lib/rqrcode/qrcode/qr_code.rb (hunk 4 above)
Comment (toncid) [inline lib/rqrcode/qrcode/qr_code.rb]: I wasn't aware you still support it. Everything below v2.0 has been deprecated horribly long.

We can use something else as the fallback value, like `'-'` or `'_'`?

**Comment by bjornblomqvist** [comment]
Good work!!! =)

But i really must insist on a test to catch regressions in the future. Currently that code path is not run during the tests so any regressions will be missed.

File: lib/rqrcode/qrcode/qr_code.rb (hunk 5 above)
Comment (toncid) [inline lib/rqrcode/qrcode/qr_code.rb]: Automatic mode recognition. Just a thought...

File: lib/rqrcode/qrcode/qr_code.rb (hunk 4 above)
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: Sure '_' will do just fine.

File: lib/rqrcode/qrcode/qr_code.rb (hunk 2 above)
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: I think just one will be best. I would like it to be :byte or :byte_8bit.

File: lib/rqrcode/qrcode/qr_code.rb (hunk 4 above)
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: You are quite right. I haven't written much during the last two years so i have missed that. Fix the above or just remove 1.8 from .travis.yml.

File: lib/rqrcode/qrcode/qr_code.rb (hunk 5 above)
Comment (bjornblomqvist) [inline lib/rqrcode/qrcode/qr_code.rb]: I think automatic is great. But I really like that you added the possibility to set mode explicit.

File: lib/rqrcode/qrcode/qr_code.rb (hunk 4 above)
Comment (toncid) [inline lib/rqrcode/qrcode/qr_code.rb]: Fixed by increasing to Ruby v1.9+

File: lib/rqrcode/qrcode/qr_code.rb (hunk 5 above)
Comment (toncid) [inline lib/rqrcode/qrcode/qr_code.rb]: OK, no changes needed then.

**Comment by toncid** [comment]
All changes applied and included a test with numeric data that I know is correct (tested the resulting QR code with a 3rd-party scanner). 

Ready for the release! :-) :+1:

**Comment by bjornblomqvist** [comment]
**Thanks for all the great work!!**

Its people like you who add what they need that makes rqrcode a little bit better year after year!

I am sorry for being slow on merging and releasing but real life got in the way. 

Best Regards
Björn Blomqvist

**Comment by toncid** [comment]
All fine! Glad to be of assistance! :+1:

**Pull request merged**